### ISSUE TITLE
Blazor: OnParametersSetAsync() of "default" layout component is executed even though page being navigated to uses a different layout component

### ISSUE THREAD

**Issue by Rauce** [aspnet/AspNetCore#13720]
### Steps to reproduce

With the following in App.razor:

```
<Router AppAssembly="typeof(Program).Assembly">
    <Found Context="routeData">
        <AuthorizeRouteView
            RouteData="routeData" 
            DefaultLayout="@typeof(MyDefaultLayout)">
        </AuthorizeRouteView>
    </Found>
</Router>
```

And a page "MyPageWithOtherLayout" which specifies an alternate layout:

```
@layout MyOtherLayout
@page "/MyPageWithOtherLayout"
```

When navigating to MyPageWithOtherLayout from a page that does use MyDefaultLayout, the logic in MyDefaultLayout.OnParametersSetAsync() is being executed. In its execution, NavigationManager.Uri is that of MyPageWithOtherLayout.

(If navigating to MyPageWithOtherLayout from another page which also uses MyOtherLayout, MyDefaultLayout.OnParametersSetAsync() is not being executed.)

### Expected behavior
MyDefaultLayout.OnParametersSetAsync() should not be executed if the page I am navigating _to_ does not use MyDefaultLayout, regardless of whether the page I'm navigating _from_ uses MyDefaultLayout or not.

### Additional context
dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview9-014004
 Commit:    8e7ef240a5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview9-014004\

Host (useful for support):
  Version: 3.0.0-preview9-19423-09
  Commit:  2be172345a

.NET Core SDKs installed:
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.1.801 [C:\Program Files\dotnet\sdk]
  2.2.202 [C:\Program Files\dotnet\sdk]
  2.2.401 [C:\Program Files\dotnet\sdk]
  3.0.100-preview8-013656 [C:\Program Files\dotnet\sdk]
  3.0.100-preview9-014004 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview8.19405.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview9.19424.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview8-28405-07 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview9-19423-09 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview8-28405-07 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview9-19423-09 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

**Comment by Rauce** [comment]
Having spent hours on testing/debugging, I have discovered that this only happens when MyDefaultLayout has this:

`[CascadingParameter] Task<AuthenticationState> AuthenticationStateTask { get; set; }`

It seems there are extra additional OnParametersSetAsync() calls with the cascading authentication state parameter, aside from the other component parameters. This may be the intended behaviour, but it makes OnParametersSetAsync() be executed twice on every navigation, which is undesirable IMO. It also causes the issue in my original post above, in which MyDefaultLayout.OnParametersSetAsync() gets executed again when navigating to a page that uses a different layout - again, undesirable.

(Is there some way inside the body of OnParametersSetAsync() to determine whether its the authentication parameter being set, or the other parameters being set?).

**Comment by SteveSandersonMS** [comment]
Thanks very much for reporting this. I've investigated and determined what's happening.

All the components are individually behaving as designed, but this particular interaction between routing and authentication leads to an undesirable consequence where some components get notified to re-render even when they are just about to be disposed.

I've written up a proposed fix for us to do here: https://github.com/aspnet/AspNetCore/issues/14566

As a quick workaround, you can manually apply the fix I'm suggesting in #14566, which is to wrap a `<CascadingAuthenticationState>` around your `App.razor` markup.

Closing this so we can track applying the fix in #14566.
